pkg/certificate/endpoint.go

```diff
@@ -0,0 +1,92 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package certificate
+
+import (
+	"context"
+
+	"go.uber.org/zap"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+
+	"storj.io/storj/pkg/certificate/authorization"
+	"storj.io/storj/pkg/identity"
+	"storj.io/storj/pkg/pb"
+)
+
+// Endpoint implements pb.CertificatesServer.
+type Endpoint struct {
+	log             *zap.Logger
+	ca              *identity.FullCertificateAuthority
+	authorizationDB *authorization.DB
+	minDifficulty   uint16
+}
+
+// NewEndpoint creates a new certificate signing gRPC server.
+func NewEndpoint(log *zap.Logger, ident *identity.FullIdentity, ca *identity.FullCertificateAuthority, authorizationDB *authorization.DB, minDifficulty uint16) *Endpoint {
+	return &Endpoint{
+		log:             log,
+		ca:              ca,
+		authorizationDB: authorizationDB,
+		minDifficulty:   minDifficulty,
+	}
+}
+
+// Sign signs the CA certificate of the remote peer's identity with the `certs.ca` certificate.
+// Returns a certificate chain consisting of the remote peer's CA followed by the CA chain.
+func (endpoint Endpoint) Sign(ctx context.Context, req *pb.SigningRequest) (_ *pb.SigningResponse, err error) {
+	defer mon.Task()(&ctx)(&err)
+	grpcPeer, ok := peer.FromContext(ctx)
+	if !ok {
+		return nil, internalErr(Error.New("unable to get peer from context"))
+	}
+
+	peerIdent, err := identity.PeerIdentityFromPeer(grpcPeer)
+	if err != nil {
+		return nil, internalErr(err)
```

Comment:
We have to log the error otherwise we lose information on what's going on.
See my comment in the `internalErr` function.

This comment applies to every `return nil, internalErr(err)` in this function.